### PR TITLE
Allow multiple images per scream

### PR DESCRIPTION
**Pull request by gfontenot**
This turned into a bit of a doozy, but this adds the ability for a single
scream to have multiple images.

This change consists of a number of interlocking parts that made it hard to
make smaller commits:

 - Images are now separate entities, with a reference back to their
  parent scream. Since we have a whole new table to play with, we're
  also going to go ahead and store more information, like the original
  filename. This is currently being used to provide some alt text, but
  might be expanded upon in the future.
- We're now performing two requests per page, instead of the one.
  We're performing one database request for screams, and another for
  the images for the screams. This mechanic works the same no matter if
  we're requesting one or many screams.
- In the case of many screams (like on the home page), we also need to
  associate images with their parent scream. This is done with a
  special helper function `joinOneToMany`. We're doing this in memory
  for the sake of simplicity.
- I've added a helper function for uploading multiple images, although
  we're not really using it in the app yet. We currently only have the
  ability to upload a single image through the UI, but it's actually
  possible to upload, store, and display multiple images with the
  provided machinery.

I've also updated my S3 stuff to use uuids as filenames, in order to
avoid name collisions moving forward. I'm not sure if there's a better
way to do this, but I'm open to ideas if this isn't the right approach.

### CODE REVIEW

File: src/Handler/Home.hs [hunk 1]
```diff
@@ -1,21 +1,27 @@
...
+import Query
 
 import Handler.ScreamDetail (singleScream)
 
 getHomeR :: Handler Html
 getHomeR = do
-    (screams, pagination) <- fetchScreams
+    (screams', pagination) <- runDB $ paginatedScreams 30 simpleWidget
+    images <- runDB $ fetchImagesForScreams screams'
+    let screams = joinOneToMany screamImage screams' images
```
Comment (tonyd256) [inline src/Handler/Home.hs]: this is so close to being something applicative-able; however, seems like the lines are too dependent on other lines to accomplish it?

What if you separated this so the join function took one scream and all the images.
`joinOneToMany :: (a -> Key b) -> [Entity a] -> Entity b` then you can use `map` like so:
`map (joinOneToMany screamImage images) screams'`

File: src/Handler/NewScream.hs [hunk 2]
```diff
@@ -27,13 +29,22 @@ data ScreamFields = ScreamFields
...
 
-fromScreamFields :: ScreamFields -> Handler Scream
-fromScreamFields f = do
+parseImages :: ScreamFields -> ScreamId -> Handler [Image]
+parseImages f sid = do
+    images <- S3.uploadImages $ maybeToList $ imageField f
+    return $ fmap createImage images
+  where
+      createImage (fileInfo, url) = Image
+          { imageScreamId = sid
```
Comment (tonyd256) [inline src/Handler/NewScream.hs]: you could leave out the record syntax if you prefer here `Image sid (fileName fileInfo) url`. Not sure you're gaining much from explicitness here.

File: src/Query.hs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    [Desc ScreamCreatedAt, Desc ScreamId] -- sort descriptors
+
+fetch404 :: ( PersistEntityBackend record ~ BaseBackend backend
+            , PersistEntity record
+            , PersistStoreWrite backend
+            , MonadIO m
+            ) => Key record -> ReaderT backend m (Entity record)
+fetch404 oid = do
+    obj <- get404 oid
+    return $ Entity oid obj
```
Comment (tonyd256) [inline src/Query.hs:41]: `Entity oid <$> get404 oid`?

File: src/Query.hs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    [ImageScreamId ==. entityKey scream]
+    []
+
+joinOneToMany :: (ToBackendKey SqlBackend a)
+              => (b -> Key a)
+              -> [Entity a]
+              -> [Entity b]
+              -> [(Entity a, [Entity b])]
+joinOneToMany f as bs = do
+    for as $ \a -> do
```
Comment (tonyd256) [inline src/Query.hs]: if you move the `for` out like my suggestion above this could be `(a, filter (belongsTo a) bs)`?

File: src/Query.hs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+fetchImagesForScreams screams =
+    selectList
+    [ImageScreamId <-. map entityKey screams]
+    []
+
+fetchImagesForScream :: ( BaseBackend backend ~ SqlBackend
+                        , MonadIO m
+                        , PersistQueryRead backend
+                        ) => Entity Scream -> ReaderT backend m [Entity Image]
+fetchImagesForScream scream =
```
Comment (tonyd256) [inline src/Query.hs]: could this call the one above with an array of 1 element for a bit more abstraction?

File: src/Handler/NewScream.hs [hunk 6]
```diff
@@ -27,13 +29,22 @@ data ScreamFields = ScreamFields
     , imageField :: Maybe FileInfo
     }
 
-fromScreamFields :: ScreamFields -> Handler Scream
-fromScreamFields f = do
+parseImages :: ScreamFields -> ScreamId -> Handler [Image]
+parseImages f sid = do
+    images <- S3.uploadImages $ maybeToList $ imageField f
+    return $ fmap createImage images
```
Comment (tonyd256) [inline src/Handler/NewScream.hs:35]: `sequence $ fmap createImage $ S3.uploadImages $ maybeToList $ imageField f`?

File: src/Handler/Home.hs (hunk 1 above)
Comment (tonyd256) [reply src/Handler/Home.hs]: This feels so close to a nicer abstraction ... maybe Pat or Joe can see something I can't.

**Review by tonyd256 — APPROVED** [review]
few comments but nothing big

File: src/Handler/Home.hs (hunk 1 above)
Comment (gfontenot) [reply src/Handler/Home.hs]: Oh yeah, I actually like this idea a lot.

File: src/Query.hs (hunk 3 above)
Comment (gfontenot) [reply src/Query.hs:41]: I had this at one point but thought this was a bit more explicit. I feel like the `get404` gets a bit lost with the inline syntax. Might leave it like this for now and revisit down the line.

File: src/Handler/NewScream.hs (hunk 6 above)
Comment (gfontenot) [reply src/Handler/NewScream.hs:35]: I dig this, but this whole thing is going to change when I add the ability to upload multiple images via the UI, so I don't know that it's worth tweaking. I'll definitely revisit this suggestion when I'm a bit further with this stuff.

**Pull request merged**